### PR TITLE
[react-dom] Allow maybe instance in findDOMNode

### PR DESCRIPTION
**Pull request by eps1lon**
Duplicate of #32667 which was closed because CI timed out. I suspect this will happen again. Pinging some collaborators for manual review. Sorry for the noise and thanks in advance.

/cc @Jessidhia @RyanCavanaugh @sandersn

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes:   
  - [v16 findDOMNode](https://github.com/facebook/react/blob/v16.0.0/src/renderers/dom/shared/findDOMNode.js#L30)
  - [v15 findDOMNode](https://github.com/facebook/react/blob/v15.5.0/src/renderers/dom/client/findDOMNode.js#L46)
  - flow maybe types include undefined and null: https://flow.org/en/docs/types/maybe/
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

File: types/react-dom/v15/index.d.ts [hunk 1]
```diff
@@ -15,7 +15,7 @@ import {
     DOMAttributes, DOMElement
 } from 'react';
 
-export function findDOMNode<E extends Element>(instance: ReactInstance): E;
+export function findDOMNode<E extends Element>(instance: ReactInstance | null | undefined): E;
```
Comment (Jessidhia) [inline types/react-dom/v15/index.d.ts]: This is one strange overload. It's one of the things the README tells you not to do (it's just a hidden `as` cast)... but it's already there 😔

File: types/react-dom/index.d.ts [hunk 2]
```diff
@@ -17,7 +17,7 @@ import {
     DOMAttributes, DOMElement, ReactNode, ReactPortal
 } from 'react';
 
-export function findDOMNode(instance: ReactInstance): Element | null | Text;
+export function findDOMNode(instance: ReactInstance | null | undefined): Element | null | Text;
```
Comment (Jessidhia) [inline types/react-dom/index.d.ts]: https://github.com/facebook/react/blob/v16.8.0/packages/react-dom/src/client/ReactDOM.js#L637

File: types/react-dom/v15/react-dom-tests.ts [hunk 3]
```diff
@@ -25,6 +25,8 @@ describe('ReactDOM', () => {
         const rootElement = document.createElement('div');
         ReactDOM.render(React.createElement('div'), rootElement);
         ReactDOM.findDOMNode(rootElement);
+        ReactDOM.findDOMNode(null);
+        ReactDOM.findDOMNode(undefined);
```
Comment (Jessidhia) [inline types/react-dom/v15/react-dom-tests.ts]: https://github.com/facebook/react/blob/v15.6.2/src/renderers/dom/client/findDOMNode.js#L44

**Review by Jessidhia — COMMENTED** [review]
This overload could be a nice-to-have (and probably what the React v15 overload actually intended to have) but probably not worth it for a deprecated API anyway.

```ts
export function findDOMNode<E extends Element>(instance: E | null | undefined): E
```

**Review by Jessidhia — APPROVED** [review]

**Comment by Jessidhia** [comment]
Errors look unrelated, either a bug in the tests or a bug in the types (but of an unrelated package).

**Comment by sandersn** [comment]
@Jessidhia yep, it's a known issue that typescript@next breaks reach-instantsearch-core. That break is tracked here: https://github.com/Microsoft/TypeScript/issues/29898

**Pull request merged**